### DISCUSSION TITLE
undefined symbol: 'luaT_typenames_' when attempting to run a lua script that requires a Pallene module.

### DISCUSSION THREAD

**Discussion by srijan-paul** [pallene-lang/pallene#362] [Q&A] [answered]
I am trying to compile a simple C command line application that uses the Lua interpreter bundled with Pallene to run some Lua scripts.

Currently, the C script invokes the Lua interpreter on a file called `app.lua`, using `luaL_dofile`.
This works just fine with both the `lua/src/lua` executable and the executable built from my C source. However, when the `app.lua` requires a Pallene module, I get the following error:

```
error loading module 'pmath' from file './pmath.so':
	./pmath.so: undefined symbol: luaT_typenames_.
```

To reproduce the problem on a simpler scale, I used the following directory structure (everything inside Pallene's root):

```
└── pallene
    ├── app.lua
    ├── pmath.so
    ├── pmath.pln
    ├── main.c
    ├── game
    ├── < pallene source >
```

Contents of `main.c`:
```c
#include "lua.h"
#include "lprefix.h"
#include "pallene_core.h"
#include "lauxlib.h"
#include "lualib.h"
#include <stdio.h>
#include <stdlib.h>

#define ASSERT(cond) \
		if (!(cond)) { \
			fprintf(stderr, "[%s %d]: assertion failed.\n", __func__, __LINE__);\
			exit(1); \
		} \

#define ERROR(...) (fprintf(stderr, __VA_ARGS__), exit(1))

int main() {
	lua_State* L = luaL_newstate();
	ASSERT(L != NULL);
	luaL_openlibs(L);
	
		
	int r = luaL_dofile(L, "./app.lua");
	if (r != LUA_OK) {
		char const* err_msg = lua_tostring(L, -1);
		fprintf(stderr, "lua interpreter exited with exitcode %d.\n", r);
		ERROR("LUA error: %s.\n", err_msg);
	}

	lua_close(L);
	return 0;
}
```

contents of `app.lua`:

```lua
local _ = require "pmath"
print "success"
```

contents of pmath.pln (irrevelant afaik):

```lua
export function add(a: integer, b: integer): integer
    return a + b
end
```

The above is compiled and run with:

```sh
gcc -I ./runtime -I ./lua/src -std=c99 -O2 -ldl -Wall -Wextra -lm -fPIC -o app main.c ./runtime/pallenelib.a ./lua/src/liblua.a
```

The compilation is successful, but then launching `./app` throws the following runtime error when dynamically loading `pmath.so`: 

```c
LUA error: error loading module 'pmath' from file './pmath.so':
	./pmath.so: undefined symbol: luaT_typenames_.
```

However running the same with `lua/src/lua` executes the script successfully.

Looking into `lua/src/Makefile`, the `LUA_T` target that builds the lua executable doesn't seem to have any different dependencies save `lua.o`.

Here is what I tried:
I tried running `nm` on all `.o` files and executables and piping the result into grep:
```sh
nm  *.o | grep "luaT_typenames"
nm  -D *.o | grep "luaT_typenames"
```
This is unable to find any object files that globally export the `luaT_typenames_`. (marked as `T`).

`luaT_typenames` is defined in the `ltm.h` header file, but I cannot seem to be able to get my executable to 'see' that definition.

Next, I tried running `ldd` on `.src/lua/lua` and `./app`.

![image](https://user-images.githubusercontent.com/50487716/115121843-72a07900-9fd2-11eb-974a-c03b4e0a1574.png)
The only `.so` that is included in the lua binary but is missing from mine is `libncurses`, which doesn't export the `luaT_typenames`.

Any help appreciated ^^

**Comment by hugomg** [comment, marked as answer]
Hmm... an interesting problem.

>  I tried running nm on all .o files and executables and piping the result into grep:

What happens if you run `nm` on the `liblua.a` and on the resulting `app` executable?

**Reply by srijan-paul** [reply]
`nm` on `liblua.a`:
```sh
$ nm ./lua/src/liblua.a | grep "luaT_typenames"
                 U luaT_typenames_
00000000000000e0 D luaT_typenames_
```
`nm` on the `app` exec:

```sh
$ nm app | grep "luaT_typenames_"
000000000003fb80 D luaT_typenames_
```

Neither export from their text segments, apparently.

**Reply by hugomg** [reply]
I've managed to reproduce the problem here. I'd like to thank you for the detailed report.

I still haven't figured out what's causing it though.

> This is unable to find any object files that globally export the luaT_typenames_. (marked as T).

IIRC, everything that's shows in the `nm` output is globally exported. For example if we run `nm` on the system Lua using `nm /usr/bin/lua` then it says that there are no exported symbols. The `T` section is for executable code (functions).

**Reply by srijan-paul** [reply]
>IIRC, everything that's shows in the nm output is globally exported.

Ah yes it does seem so according to the  [man pages](https://man7.org/linux/man-pages/man1/nm.1p.html#STDOUT).
I did try linking with the individual object files too instead of the `liblua.a` but it didn't help.

Could it be that the shared object file that is emitted by the Pallene compiler needs to be built with that exported symbol?
I tried invoking gcc manually after using the `--emit-c` flag, but the compiler threw an error complaining about relocation against `luaT_typenames` in the shared object file.
Compiling again with `-fPIC` didn't seem to help either.

**Reply by hugomg** [reply]
Looks like you were doing the right thing when you ran `nm -D`. Apparently there's a distinction between "dynamic symbols" and "normal symbols". Only the symbols that appear in the `nm -D` list are visible to dynamically-linked libraries.

https://stackoverflow.com/questions/15345543/when-to-use-dynamic-option-in-nm

If we run `nm -D lua/src/lua`, the `luaT_typenames_` is there but if we run `nm -D app` it's not. Perhaps it the culprit could be our compilation flags or something like that?

**Reply by hugomg** [reply]
Aha! I think I figured out what it is. On Linux we need to use the `-E` or `--export-dynamic` linker flag.
https://stackoverflow.com/questions/26294841/having-object-file-symbols-become-dynamic-symbols-in-executable

In the Lua makefile, it uses this flag when linking the `lua` executable.

```
linux-noreadline:
	$(MAKE) $(ALL) SYSCFLAGS="-DLUA_USE_LINUX" SYSLIBS="-Wl,-E -ldl"

linux-readline:
	$(MAKE) $(ALL) SYSCFLAGS="-DLUA_USE_LINUX -DLUA_USE_READLINE" SYSLIBS="-Wl,-E -ldl -lreadline"
```

**Reply by srijan-paul** [reply]
That's great!
I've been trying a couple of things but this doesn't seem to be working for me still.

I'm compiling like so:
```
gcc -Wall -std=c99 -I ./runtime -I ./lua/src -c  main.c
```
and linking:
```
gcc -o app -I ./runtime -I ./lua/src -export-dynamic -lm -ldl main.o  ./lua/src/liblua.a ./runtime/pallenelib.a
```
However it still croaks the very same error for me.
I also tried the `-rdynamic` gcc flag that passes `-E` to the linker, but in vain.

I tried linking with `ld` using similar options:
```
ld -o app -I ./runtime -I ./lua/src -E main.o -lm -ldl ./lua/src/liblua.a ./runtime/pallenelib.a
```
However this time I had the following errors:
```
ld: ./lua/src/liblua.a(lauxlib.o): undefined reference to symbol 'getc@@GLIBC_2.2.5'
ld: /usr/lib/libc.so.6: error adding symbols: DSO missing from command line
```
I'm not sure whats causing this. Could it be because of incompatible versions of `libc`?
`getc` does not require any explicit library args afaik, so it could be the version in the identifier `2.2.5`?

**Reply by hugomg** [reply]
These compilation lines worked for me:

    gcc -I ./runtime -I ./lua/src -std=c99 -O2 -Wall -Wextra -fPIC -c -o main.o main.c
    gcc -ldl -lm -Wl,-E -o app main.o ./runtime/pallenelib.a ./lua/src/liblua.a

The `-Wl,-E` part passes the `-E` flag to the linker.

> I'm not sure whats causing this. Could it be because of incompatible versions of libc?

Dunno. I'm using Fedora 33, if that helps.

**Reply by srijan-paul** [reply]
Awesome, that worked!
Thank you very much ^^
